src/Mvc/Mvc.ViewFeatures/src/Rendering/TagBuilder.cs

```diff
@@ -154,51 +160,39 @@ public static string CreateSanitizedId(string? name, string invalidCharReplaceme
         }
 
         // If there are no invalid characters in the string, then we don't have to create the buffer.
-        var firstIndexOfInvalidCharacter = 1;
-        for (; firstIndexOfInvalidCharacter < name.Length; firstIndexOfInvalidCharacter++)
+        var indexOfInvalidCharacter = name.AsSpan(1).IndexOfAnyExcept(_html401IdChars);
+        var firstChar = name[0];
+        var startsWithAsciiLetter = char.IsAsciiLetter(firstChar);
+        if (startsWithAsciiLetter && indexOfInvalidCharacter < 0)
         {
-            if (!Html401IdUtil.IsValidIdCharacter(name[firstIndexOfInvalidCharacter]))
-            {
-                break;
-            }
+            return name;
         }
 
-        var firstChar = name[0];
-        var startsWithAsciiLetter = char.IsAsciiLetter(firstChar);
         if (!startsWithAsciiLetter)
         {
             // The first character must be a letter according to the HTML 4.01 specification.
             firstChar = 'z';
         }
 
-        if (firstIndexOfInvalidCharacter == name.Length && startsWithAsciiLetter)
-        {
-            return name;
-        }
-
         var stringBuffer = new StringBuilder(name.Length);
```

Comment:
Maybe this `StringBuilder`, could be replaced with an `ArrayPool<char>` to avoid some allocations, but `invalidCharReplacement` is string, so calculating the rental size etc. probably not worth the effort for the extra complexity if not on hot path (and I don't think so).